src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,205 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+
+pub const DEFAULT_BUF_SIZE: usize = 64 * 1024;
+
+
+pub fn read_to_end_zeroed<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> io::Result<usize> {
+    unsafe { read_to_end(r, buf, true) }
+}
+
+pub unsafe fn read_to_end_uninitialized<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>)
+                                                          -> io::Result<usize> {
+    read_to_end(r, buf, false)
+}
+
+// Used to fill a Vec from a Read type until there's no more data to come.
+// Uses an adaptive system to extend the vector when it fills. We want to
+// avoid paying to allocate and zero a huge chunk of memory if the reader only
+// has 4 bytes while still making large reads if the reader does have a ton
+// of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
+// time is 4,500 times (!) slower than this if the reader has a very small
+// amount of data to return.
+// function is unsafe because it can (optionally) work with uninitialised
+// buffers by passing zero_buf=false.  If using zero_buf=false, users should
+// ensure that their Read implementation does not read data from the buffer,
+// and always returns the correct number of bytes read.  Otherwise, behaviour
+// is undefined.
+unsafe fn read_to_end<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>, zero_buf: bool)
+                                        -> io::Result<usize> {
+    let start_len = buf.len();
+    let mut len = start_len;
+    let mut new_write_size = 16;
+    let ret;
+    loop {
+        if len == buf.len() {
+            if new_write_size < DEFAULT_BUF_SIZE {
+                new_write_size *= 2;
+            }
+            if zero_buf {
+                buf.resize(len + new_write_size, 0);
+            } else {
+                buf.reserve(new_write_size);
+                buf.set_len(len + new_write_size);
+            }
+        }
+
+        match r.read(&mut buf[len..]) {
+            Ok(0) => {
+                ret = Ok(len - start_len);
+                break;
+            }
+            Ok(n) => len += n,
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => { }
+            Err(e) => {
+                ret = Err(e);
+                break;
+            }
+        }
+    }
+
+    buf.truncate(len);
+    ret
+}
+
+#[cfg(test)]
+mod tests {
+    use prelude::v1::*;
+    use io::prelude::*;
+    use super::*;
+    use io::ErrorKind;
+    use io;
+    use test;
+
+    struct LimitedRepeat {
+        byte: u8,
+        limit: usize,
+        count: usize
+    }
+
+    impl LimitedRepeat {
+        fn new(byte: u8, limit: usize) -> LimitedRepeat {
+            LimitedRepeat { byte: byte, limit: limit, count: 0}
+        }
+
+        fn exhausted(&self) -> bool {
+            self.count == self.limit
+        }
+    }
+
+    impl Read for LimitedRepeat {
```

Comment:
Good call!  Since io::utils is private, I've pulled the code into sys_common::io, and then reexported it in io::utils.